### DISCUSSION TITLE
About the L2 level operation of recidive

### DISCUSSION THREAD

**Discussion by nolix819** [fail2ban/fail2ban#3769] [Q&A] [answered]
About the L2 level operation of recidive

O.S. Rocky9.3
fail2ban-selinux-1.0.2-12.el9.noarch
fail2ban-server-1.0.2-12.el9.noarch
fail2ban-firewalld-1.0.2-12.el9.noarch
fail2ban-sendmail-1.0.2-12.el9.noarch
fail2ban-1.0.2-12.el9.noarch

jail.local
```
[postfix_L1]
enabled = true
filter = postfix
mode = aggressive
port = 25
#logpath = %(postfix_log)s
logpath = /var/log/maillog*
backend = %(postfix_backend)s
banaction = iptables-multiport-tcpreset
findtime = 300
bantime = 300
maxretry = 4
[postfix_L2]
enabled = true
filter = postfix_L2
port = 25,465,587,110,995,143,993,4190,65110
logpath = /var/log/fail2ban.log
banaction = iptables-multiport-tcpreset
bantime.increment = false
bantime.rndtime =
bantime.maxtime =
findtime = 1d
bantime = 1d
maxretry = 4
```
filter.d/postfix.conf is contributed one.
filter.d/postfix_L2.conf
```
[INCLUDES]
# Read common prefixes. If any customizations available -- read them from
# common.local
before = common.conf

[Definition]
_daemon = (?:fail2ban(?:-server|\.actions)\s*)

# The name of the jail that this filter is used for. In jail.conf, name the jail using
# This filter 'recidive', or supply another name with `filter = recidive[_jailname="jail"]`
_jailname = postfix_L1

#failregex = ^%(__prefix_line)s(?:\s*fail2ban\.actions\s*%(__pid_re)s?:\s+)?NOTICE\s+\[(?!%(_jailname)s\])(?:.*)\]\s+Ban\s+<HOST>\s*$
failregex = .*fail2ban.actions.*NOTICE.*postfix_L1.*Ban+<HOST>
datepattern = ^{DATE}
ignoreregex =
journalmatch = _SYSTEMD_UNIT=fail2ban.service PRIORITY=5

# Author: Tom Hendrikx, modifications by Amir Caspi
```
```
2024-06-20 08:21:34,705 fail2ban.actions [796727]: NOTICE [postfix_L1] Ban 121.229.2.220
2024-06-20 08:26:36,296 fail2ban.actions [796727]: NOTICE [postfix_L1] Ban 121.229.2.220
2024-06-20 08:31:37,093 fail2ban.actions [796727]: NOTICE [postfix_L1] Ban 121.229.2.220
2024-06-20 08:36:38,071 fail2ban.actions[796727]: NOTICE [postfix_L1] Ban 121.229.2.220
2024-06-20 08:41:40,465 fail2ban.actions[796727]: NOTICE [postfix_L1] Ban 121.229.2.220
2024-06-20 08:46:44,056 fail2ban.actions[796727]: NOTICE [postfix_L1] Ban 121.229.2.220
2024-06-20 08:51:44,847 fail2ban.actions[796727]: NOTICE [postfix_L1] Ban 121.229.2.220
2024-06-20 08:56:46,444 fail2ban.actions[796727]: NOTICE [postfix_L1] Ban 121.229.2.220
2024-06-20 09:01:48,847 fail2ban.actions[796727]: NOTICE [postfix_L1] Ban 121.229.2.220
2024-06-20 09:06:49,647 fail2ban.actions[796727]: NOTICE [postfix_L1] Ban 121.229.2.220
2024-06-20 09:11:50,436 fail2ban.actions[796727]: NOTICE [postfix_L1] Ban 121.229.2.220
2024-06-20 09:16:51,228 fail2ban.actions [796727]: NOTICE [postfix_L1] Ban 121.229.2.220
2024-06-20 09:21:52,827 fail2ban.actions [796727]: NOTICE [postfix_L1] Ban 121.229.2.220
2024-06-20 09:26:57,029 fail2ban.actions [796727]: NOTICE [postfix_L1] Ban 121.229.2.220
```
As shown above, postfix_L1 is working properly, but reciprocal L2 is not working properly.
Regex didn't work so I changed it to a simpler one.
If you notice anything, I'd appreciate it if you could let me know.

Noriyuki Hayashi

**Comment by sebres** [comment, marked as answer]
It looks like we have a small inconsistence by the comment in recidive filter:
https://github.com/fail2ban/fail2ban/blob/ab9d41e5309b417a3c7a84fa8f03cf4f93831f1b/config/filter.d/recidive.conf#L27
(and in your filter in the same way too)

The issue is, `(?!%(_jailname)s\])` is negative lookahead and will match only if not followed by specified jailname.
So the comment is lying at the moment (and one'd specify an exclusion rather).

The fix could be simple:
```diff
- _jailname = recidive
+ _jailname = (?!recidive\])[^\]]*

- failregex = ^%(__prefix_line)s(?:\s*fail2ban\.actions\s*%(__pid_re)s?:\s+)?NOTICE\s+\[(?!%(_jailname)s\])(?:.*)\]\s+Ban\s+<HOST>\s*$
+ failregex = ^%(__prefix_line)s(?:\s*fail2ban\.actions\s*%(__pid_re)s?:\s+)?NOTICE\s+\[<_jailname>\]\s+Ban\s+<HOST>
```

As for the rest...

1. approach with recidive jail is a rudiment - normally, `bantime.increment = true` is fully enough as a replacement for that.
2. however recidive jail (and filter) were made to ban recidive evildoers from **all jails** not from single jail only.
3. you don't need a new filter, you can use stock [recidive filter](https://github.com/fail2ban/fail2ban/blob/master/config/filter.d/recidive.conf) instead, and use it in your jail:
```ini
[postfix_L2]
# for all jails:
filter = recidive
# or for single jail (after fix):
filter = recidive[_jailname=postfix_L1]
...
```
4. Alternatively, you can use something like that (with direct RE without filter):
```ini
[postfix_L2]
filter = 
failregex = ^\s*(?:fail2ban(?:-server|\.actions)\s*)\[\d+\]: NOTICE \[postfix_L1\] Ban <ADDR>
...
```

**Reply by nolix819** [reply]
Hello Mr.Sebres,

Thank you so much.
I tried to change that I am confirming now.
I will report after I got success with the configuration for other people.

Kind regards,

**Comment by sebres** [comment]
Fixed in a7f3a04b0e0c50a82db0b7a27cc783fd5faad0ec

**Reply by nolix819** [reply]
Hello Sebres,

Thank you for your great teaching and I report for other people.

/etc/fail2ban/Jail.local   (Picked up only apache-access)  maxretry is adjusted for the test.
[apache-access_L1]
enabled   = true
filter    = apache-access
port      = 80,443
logpath   = %(apache_access_log)s
backend   = polling
banaction = iptables-multiport-tcpreset
findtime  = 300
bantime   = 300
maxretry  = 2

[apache-access_L2]
enabled   = true
filter    = recidive[_jailname="apache-access_L1"]
port      = 80,443
logpath   = /var/log/fail2ban.log
banaction = iptables-multiport-tcpreset
bantime.increment = false
bantime.rndtime =
bantime.maxtime =
findtime  = 1d
bantime   = 1d
maxretry  = 2

[apache-access_L3]
enabled   = true
filter    = recidive[_jailname="apache-access_L2"]
port      = 80,443
logpath   = /var/log/fail2ban.log
banaction = ipset-TCP
bantime.increment = false
bantime.rndtime =
bantime.maxtime =
findtime  = 4w
bantime   = -1
maxretry  = 4

 /etc/fail2ban/action.d/iptables-multiport-tcpreset.conf (multiport with tcp-reset version)
# Fail2Ban configuration file
# Author: Noriyuki Hayashi <wats@wats.gr.jp>
# Obsolete: superseded by iptables[type=multiport]
[INCLUDES]
before = iptables.conf
[Definition]
type = multiport
actionban = <iptables> -I f2b-<name> -p <protocol> -s <ip> -m multiport --dports <port> -j <blocktype>
actionunban = <iptables> -D f2b-<name> -p <protocol> -s <ip> -m multiport --dports <port> -j <blocktype>
[Init]
blocktype = REJECT --reject-with tcp-reset
[Init?family=inet6]
blocktype = REJECT --reject-with tcp-reset

 /etc/fail2ban/action.d/ipset-TCP.conf (This action is for IPSET)
# Fail2Ban configuration file
# Author: Cyril Jaquier
# Modified by Yaroslav Halchenko for multiport banning
# Modofied: Noriyuki Hayashi <wats@wats.gr.jp>
# Obsolete: superseded by iptables[type=multiport]
[INCLUDES]

[Definition]
actionban = ipset -q add deny.LTCP <ip>;echo <ip> >> /etc/sysconfig/deny.LTCP
###actionunban = ipset -q del deny.LTCP <ip>

/etc/fail2ban/filter.d/apache-access.conf (apache access_log combined version)
# Fail2Ban configuration file
# Author: https://wats.gr.jp/
# $Revision: 1.0 $
# apache-40x, apache-attack, apache-badbot, apache-pass, php-url-fopen
# apache-directory traversal, apache-reverse-proxy, apache-wordpress
[INCLUDES]
# overwrite with apache-common.local if _apache_error_client is incorrect.
before = apache-common.conf

[Definition]
badbotscustom = EmailCollector|WebEMailExtrac|TrackBack/1\.02|sogou music spider|BLEXBot|ltx71|DotBot|Barkrowler
badbots = Atomic_Email_Hunter/4\.0|atSpider/1\.0|autoemailspider|bwh3_user_agent|China Local Browse 2\.6|ContactBot/0\.2|ContentSmartz|DataCha0s/2\.0|DBrowse 1\.4b|DBrowse 1\.4d|Demo Bot DOT 16b|Demo Bot Z 16b|DSurf15a 01|DSurf15a 71|DSurf15a 81|DSurf15a VA|EBrowse 1\.4b|Educate Search VxB|EmailSiphon|EmailSpider|EmailWolf 1\.00|ESurf15a 15|ExtractorPro|Franklin Locator 1\.8|FSurf15a 01|Full Web Bot 0416B|Full Web Bot 0516B|Full Web Bot 2816B|8484 Boston Project v 1\.0|Atomic_Email_Hunter/4\.0|atSpider/1\.0|autoemailspider|bwh3_user_agent|China Local Browse 2\.6|ContactBot/0\.2|ContentSmartz|DataCha0s/2\.0|DBrowse 1\.4b|DBrowse 1\.4d|Demo Bot DOT 16b|Demo Bot Z 16b|DSurf15a 01|DSurf15a 71|DSurf15a 81|DSurf15a VA|EBrowse 1\.4b|Educate Search VxB|EmailSiphon|EmailSpider|EmailWolf 1\.00|ESurf15a 15|ExtractorPro|Franklin Locator 1\.8|FSurf15a 01|Full Web Bot 0416B|Full Web Bot 0516B|Full Web Bot 2816B|8484 Boston Project v 1\.0|Atomic_Email_Hunter/4\.0|atSpider/1\.0|autoemailspider|bwh3_user_agent|China Local Browse 2\.6|ContactBot/0\.2|ContentSmartz|DataCha0s/2\.0|DBrowse 1\.4b|DBrowse 1\.4d|Demo Bot DOT 16b|Demo Bot Z 16b|DSurf15a 01|DSurf15a 71|DSurf15a 81|DSurf15a VA|EBrowse 1\.4b|Educate Search VxB|EmailSiphon|EmailSpider|EmailWolf 1\.00|ESurf15a 15|ExtractorPro|Franklin Locator 1\.8|FSurf15a 01|Full Web Bot 0416B|Full Web Bot 0516B|Full Web Bot 2816B|8484 Boston Project v 1\.0|Atomic_Email_Hunter/4\.0|atSpider/1\.0|autoemailspider|bwh3_user_agent|China Local Browse 2\.6|ContactBot/0\.2|ContentSmartz|DataCha0s/2\.0|DBrowse 1\.4b|DBrowse 1\.4d|Demo Bot DOT 16b|Demo Bot Z 16b|DSurf15a 01|DSurf15a 71|DSurf15a 81|DSurf15a VA|EBrowse 1\.4b|Educate Search VxB|EmailSiphon|EmailSpider|EmailWolf 1\.00|ESurf15a 15|ExtractorPro|Franklin Locator 1\.8|FSurf15a 01|Full Web Bot 0416B|Full Web Bot 0516B|Full Web Bot 2816B|8484 Boston Project v 1\.0|Atomic_Email_Hunter/4\.0|atSpider/1\.0|autoemailspider|bwh3_user_agent|China Local Browse 2\.6|ContactBot/0\.2|ContentSmartz|DataCha0s/2\.0|DBrowse 1\.4b|DBrowse 1\.4d|Demo Bot DOT 16b|Demo Bot Z 16b|DSurf15a 01|DSurf15a 71|DSurf15a 81|DSurf15a VA|EBrowse 1\.4b|Educate Search VxB|EmailSiphon|EmailSpider|EmailWolf 1\.00|ESurf15a 15|ExtractorPro|Franklin Locator 1\.8|FSurf15a 01|Full Web Bot 0416B|Full Web Bot 0516B|Full Web Bot 2816B

### apache-40x
failregex = ^<HOST> - .* "(GET|POST|HEAD).*HTTP.*" (400|401|403|404|405|408|418|444) .*$
### apache-attack
            ^<HOST> -.*"CONNECT .*"$
            ^<HOST> - - (?:\[[^]]*\] )+\"(GET|POST) /(?i)(PMA|phptest|phpmyadmin|myadmin|mysql|mysqladmin|sqladmin|mypma|admin|xampp|mysqldb|mydb|db|pmadb|phpmyadmin1|phpmyadmin2|cgi-bin|phpunit|webuploader|python-requests)
### apache-badbot
            ^<HOST> -.*"(GET|POST).*HTTP.*"(?:%(badbots)s|%(badbotscustom)s)"$
### apache-pass (knocking request)
            ^<HOST> - \w+ \[\] "GET <knocking_url> HTTP/1\.[01]" 200 \d+ ".*" "[^-].*"$
### indication of a fopen url php injection
            ^<HOST> -.*"(GET|POST).*\?.*\=http\:\/\/.* HTTP\/.*$
### directory traversal
            ^<HOST>.*] "GET /.*\.\./\.\./\.\./
### Transmission with Apache2 Reverse Proxy
            ^<HOST> - - .*transmission.*HTTP/[0-9]+(.[0-9]+)?" 401
### wordpress login brute forcing
            ^<HOST> - .* "(GET | POST)/ + wp-login.php
            ^<HOST> - .* "(GET | POST)/ + xmlrpc.php

# Option: ignoreregex
# Notes.: regex to ignore. If this regex matches, the line is ignored.
# Values: TEXT
ignoreregex = \.(?i)(jpe?g|gif|png|bmp|pdf|js|css|woff|eot|ttf|ico|txt|xml|swf|xlsx?|docx?|pptx?)
              ^<HOST> -.*"HEAD .* 200.*"$

datepattern = ^[^\[]*\[({DATE})
              {^LN-BEG}
[Init]
knocking_url = /knocking/

/etc/fail2ban/filter.d/recidive.conf (This filter is for apache-access_L2 and apache-access_L3)
# Fail2Ban filter for repeat bans
#
# This filter monitors the fail2ban log file, and enables you to add long
# time bans for ip addresses that get banned by fail2ban multiple times.
#
# Reasons to use this: block very persistent attackers for a longer time,
# stop receiving email notifications about the same attacker over and
# over again.
#
# This jail is only useful if you set the 'findtime' and 'bantime' parameters
# in jail.conf to a higher value than the other jails. Also, this jail has its
# drawbacks, namely in that it works only with iptables, or if you use a
# different blocking mechanism for this jail versus others (e.g. hostsdeny
# for most jails, and shorewall for this one).

[INCLUDES]
# Read common prefixes. If any customizations available -- read them from
# common.local
before = common.conf

[Definition]
_daemon = (?:fail2ban(?:-server|\.actions)\s*)

# The name of the jail that this filter is used for. In jail.conf, name the jail using
# this filter 'recidive', or supply another name with `filter = recidive[_jailname="jail"]`
_jailname = (?!recidive\])[^\]]*

failregex = ^%(__prefix_line)s(?:\s*fail2ban\.actions\s*%(__pid_re)s?:\s+)?NOTICE\s+\[<_jailname>\]\s+Ban\s+<HOST>
datepattern = ^{DATE}
ignoreregex =
journalmatch = _SYSTEMD_UNIT=fail2ban.service PRIORITY=5
# Author: Tom Hendrikx, modifications by Amir Caspi

Test Result
# less /var/log/fail2ban.log | grep 5.183.103.248
2024-06-25 15:13:36,511 fail2ban.filter         [1158006]: INFO    [apache-ddos_L1] Found 5.183.103.248 - 2024-06-25 15:13:36
2024-06-25 15:13:37,022 fail2ban.filter         [1158006]: INFO    [apache-access_L1] Found 5.183.103.248 - 2024-06-25 15:13:36
2024-06-25 15:13:37,623 fail2ban.filter         [1158006]: INFO    [apache-access_L1] Found 5.183.103.248 - 2024-06-25 15:13:37
2024-06-25 15:13:38,315 fail2ban.filter         [1158006]: INFO    [apache-ddos_L1] Found 5.183.103.248 - 2024-06-25 15:13:38
2024-06-25 15:13:38,825 fail2ban.filter         [1158006]: INFO    [apache-access_L1] Found 5.183.103.248 - 2024-06-25 15:13:38
2024-06-25 15:13:39,426 fail2ban.filter         [1158006]: INFO    [apache-access_L1] Found 5.183.103.248 - 2024-06-25 15:13:39
2024-06-25 15:13:39,444 fail2ban.actions        [1158006]: NOTICE  [apache-access_L1] Ban 5.183.103.248
2024-06-25 15:13:39,718 fail2ban.filter         [1158006]: INFO    [apache-access_L2] Found 5.183.103.248 - 2024-06-25 15:13:39
2024-06-25 15:18:39,015 fail2ban.actions        [1158006]: NOTICE  [apache-access_L1] Unban 5.183.103.248
2024-06-25 21:49:03,152 fail2ban.filter         [1227503]: INFO    [apache-access_L2] Found 5.183.103.248 - 2024-06-25 15:13:39
2024-06-26 01:03:52,982 fail2ban.filter         [1227503]: INFO    [apache-ddos_L1] Found 5.183.103.248 - 2024-06-26 01:03:52
2024-06-26 01:03:53,617 fail2ban.filter         [1227503]: INFO    [apache-access_L1] Found 5.183.103.248 - 2024-06-26 01:03:53
2024-06-26 01:03:54,218 fail2ban.filter         [1227503]: INFO    [apache-access_L1] Found 5.183.103.248 - 2024-06-26 01:03:54
2024-06-26 01:03:54,418 fail2ban.actions        [1227503]: NOTICE  [apache-access_L1] Ban 5.183.103.248
2024-06-26 01:03:54,448 fail2ban.filter         [1227503]: INFO    [apache-access_L2] Found 5.183.103.248 - 2024-06-26 01:03:54
2024-06-26 01:03:54,971 fail2ban.actions        [1227503]: NOTICE  [apache-access_L2] Ban 5.183.103.248
2024-06-26 01:03:55,067 fail2ban.filter         [1227503]: INFO    [apache-access_L3] Found 5.183.103.248 - 2024-06-26 01:03:54
2024-06-26 01:08:54,798 fail2ban.actions        [1227503]: NOTICE  [apache-access_L1] Unban 5.183.103.248

It works as expected.
At the moment, I can't imagine operating without fail2ban.
It's a great piece of software.

Kind regards,
Noriyuki Hayashi
